spec/support/helpers.rb

```diff
@@ -0,0 +1,142 @@
+require "tmpdir"
+require "open3"
+
+module Spec
+  module Helpers
+    def self.bundler_version
+      return @bundler_version if defined?(@bundler_version)
+      versions = Bundler.with_clean_env do
+        `gem list bundler | grep "bundler "`.chomp.scan(/\d+\.\d+\.\d+/)
+      end
+      to_find = ENV["TEST_BUNDLER_VERSION"] || ENV["BUNDLER_VERSION"]
+      @bundler_version = versions.detect { |v| v.include?(to_find) }
+      raise "Unable to find bundler version: #{to_find.inspect}" if @bundler_version.nil?
+      @bundler_version
+    end
+
+    def self.bundler_short_version
+      bundler_version.rpartition(".").first
+    end
+
+    def self.bundler_cli_version
+      "_#{bundler_version}_"
+    end
+
+    def self.global_bundler_d_dir
+      @global_bundler_d_dir ||= Pathname.new("~/.bundler.d").expand_path
+    end
+
+    def self.global_bundler_d_backup_dir
+      @global_bundler_d_backup_dir ||= Pathname.new("~/.bundler.d.rspec_backup").expand_path
+    end
+
+    def self.backup_global_bundler_d
+      FileUtils.rm_rf(global_bundler_d_backup_dir)
+      FileUtils.mv(global_bundler_d_dir, global_bundler_d_backup_dir) if global_bundler_d_dir.exist?
+    end
+
+    def self.restore_global_bundler_d
+      FileUtils.rm_rf(global_bundler_d_dir)
+      FileUtils.mv(global_bundler_d_backup_dir, global_bundler_d_dir) if global_bundler_d_backup_dir.exist?
+    end
+
+    attr_reader :out, :err, :process_status
+
+    def bundler_version
+      Helpers.bundler_version
+    end
+
+    def bundler_short_version
+      Helpers.bundler_short_version
+    end
+
+    def app_dir
+      @app_dir ||= Pathname.new(Dir.mktmpdir)
+    end
+
+    def rm_app_dir
+      return unless @app_dir
+      FileUtils.rm_rf(@app_dir)
+      @app_dir = nil
+    end
+
+    def rm_global_bundler_d_dir
+      FileUtils.rm_rf(Helpers.global_bundler_d_dir)
+    end
+
+    def with_path_based_gem(source_repo)
+      Dir.mktmpdir do |path|
+        path = Pathname.new(path)
+        Dir.chdir(path) { `git clone --depth 1 #{source_repo} the_gem 2>/dev/null` }
+        path = path.join("the_gem")
+
+        yield path
+      end
+    end
+
+    def write_gemfile(contents)
+      File.write(app_dir.join("Gemfile"), contents)
+    end
+
+    def lockfile
+      file = app_dir.join("Gemfile.lock")
+      Bundler::LockfileParser.new(file.read) if file.exist?
+    end
+
+    def lockfile_specs
+      return unless (lf = lockfile)
+      lf.specs.map { |s| [s.name, s.version.to_s] }
+    end
+
+    def raw_bundle(command, verbose: false)
+      command = "bundle #{Helpers.bundler_cli_version} #{command} #{"--verbose" if verbose}".strip
+      out, err, process_status = Bundler.with_clean_env do
+        Open3.capture3(command, :chdir => app_dir)
+      end
+      return command, out, err, process_status
```

Comment:
Can we simplify this to not return anything?  Or at least only the command?

`out`, `err`, and `process_status` are each instance variables as well, so it seems silly to set them twice (effectively) instead of just making them `attr_accessor`s.

Reply:
I have split `raw_bundle` from `bundle` on purpose.  The raw method just runs and returns the output....it's not setting the instance vars, those are local vars.  Separately, the `bundle` method takes the results of `raw_bundle`, sets the instance vars, and does a basic exitstatus check.

The `raw_bundle` is useful as a separate method because I can run bundler commands independent of the command-under-test, and not affect the instance vars.  In particular I need this to get at the rack gem's location when we clone from git.

Reply:
Alright, fair enough.  👍